onnxruntime/core/providers/cpu/tensor/col2im.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include "core/providers/cpu/tensor/col2im.h"
+
+#include "core/framework/element_type_lists.h"
+#include "core/framework/TensorSeq.h"
+#include "core/providers/common.h"
+#include "core/framework/copy.h"
+#include "core/common/safeint.h"
+#include "core/providers/op_kernel_type_control.h"
+
+namespace onnxruntime {
+
+#define REGISTER_KERNEL_TYPED(T)                                                            \
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                                           \
+      Col2Im,                                                                               \
+      1,                                                                                    \
+      T,                                                                                    \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()), \
+      Col2Im<T>);
+
+REGISTER_KERNEL_TYPED(float)
+
+template <typename T>
+Status Col2Im<T>::Compute(OpKernelContext* context) const {
+  const auto* col_input = context->Input<Tensor>(0);
+  const auto* image_shape = context->Input<Tensor>(1);
+  const auto* kernel_shape = context->Input<Tensor>(2);
+  std::cout << "Status Col2Im<T>::Compute(OpKernelContext* context)" << std::endl;
+
+  const T* col_input_data = col_input->template Data<T>();
+  TensorShape col_input_shape = col_input->Shape();
+  int64_t col_input_C = col_input_shape[1];
+  const auto col_input_N = col_input_shape[0];
+
+  int64_t image_shape_size = 1;
+  int64_t kernel_shape_size = 1;
+  int64_t kernel_shape_rank = 0;
+  for (auto i=0; i < image_shape->Shape().Size(); ++i) {
+    ++kernel_shape_rank;
+    image_shape_size *=  image_shape->Data<int64_t>()[i];
+    kernel_shape_size *=  kernel_shape->Data<int64_t>()[i];
+    // col_input_C computed as => (C*n-ary-prod{kernel_shape}) / n-ary-prod{kernel_shape}
+    col_input_C /= kernel_shape->Data<int64_t>()[i];
```

Comment:
nit: suggest eliminating this, and replacing this with
```cc
   const int64_t C = col_input_c / kernel_shape_size
```
after the loop. It helps to have different names like `col_input_c` and `C` to distinguish them.

Reply:
So, does this mean that the `kernel_shape` already incorporates the effect of dilation?

Reply:
I think `kernel_shape` does not already incorporage dilation. Eg, look at [this line](https://github.com/microsoft/onnxruntime/blob/f92e47e95b13a240e37caf7b36577983544f98fc/onnxruntime/core/util/math_cpu.cc#L755).  So, I think adjustments are required for that.

Reply:
> nit: suggest eliminating this, and replacing this with
> 
> ```c++
>    const int64_t C = col_input_c / kernel_shape_size
> ```
> 
> after the loop. It helps to have different names like `col_input_c` and `C` to distinguish them.

I didnt follow this comment. Should I remove from `43-46`? The suggested new code uses the suggested removal. `image_shape_size` is used in other places too

Reply:
No, I meant removing only line 46. Main suggestion is to give different names to `col_input_shape[1]` and `col_input_shape[1] / kernel_shape_size`. Otherwise,  the name `col_input_C` is confusing.

Reply:
And `C` would be used where `col_input_C` is used now. This is just a renaming suggestion. (Moving the line out of the loop is just a simplification that goes along with it.)

Reply:
done

Reply:
@gramalingam I have added a `dilated_kernel_shape`. Will create a unit test with dilation later